### DISCUSSION TITLE
Hiren's BootCD

### DISCUSSION THREAD

**Discussion by ghost** [netbootxyz/netboot.xyz#46] [Ideas]
Is it possible to add Hiren's BootCD (http://www.hirensbootcd.org/)? It boots fine with the unzipped iso but not the zip directly from the site itself.

**Comment by antonym** [comment]
The zip file didn't seem to play with memdisk nicely and it doesn't look like there's been development on the tool in a while.  We could look into it if there was a hosted ISO.

**Comment by Enrico204** [comment]
The ISO seems to be available now, and the development has restored, so maybe this issue can be reopened @antonym ?

PS: direct ISO URL: https://www.hirensbootcd.org/files/HBCD_PE_x64.iso

**Comment by Sander0542** [comment]
I would love to see this in Netboot

**Comment by antonym** [comment]
Gave this a whirl a week ago, dumping notes here, appears that wimboot was looking for bootmgr.exe which was not present in the actual wim.  I think we could get it working but would need that bootloader present.  The bootmgr in the root of the ISO is not the one we would need either.

If anyone has cycles to dig in, it would be appreciated.

```
#!ipxe

set win_base_url https://github.com/netbootxyz/asset-mirror/releases/download/v1.0.1-0570770e/
imgfree
kernel https://boot.netboot.xyz/wimboot
initrd -n bootmgr ${win_base_url}/bootmgr       bootmgr ||
initrd -n bootmgr.efi ${win_base_url}/bootmgr.efi   bootmgr.efi ||      
initrd -n bcd         ${win_base_url}/BCD      bcd ||
initrd -n boot.sdi    ${win_base_url}boot.sdi boot.sdi ||
initrd -n boot.wim    ${win_base_url}boot.wim boot.wim
```

**Comment by lrdfrd** [comment]
any progress on this? been trying to get this to work as well and pointing to the supplied bootmgr and bootx64 doesn't seem to work, 
cant wimboot load ISOs? why wouldn't we aim it at that? (been trying myself, haven't figured out the syntax)

**Comment by FM1337** [comment]
```
#!ipxe

set base-url <location-to-folder-containing-extracted-iso>

:start
menu Please choose an option:
item hirens Hirens Boot CD PE x64
item exit Exit
choose --default hirens --timeout 5000 selected
goto ${selected}

:hirens
echo Booting Hirens Boot CD PE x64...
imgfree
kernel https://boot.netboot.xyz/wimboot
initrd -n bootmgr ${base-url}/BOOTMGR       bootmgr ||
initrd -n bootmgr.efi ${base-url}/bootmgr.efi   bootmgr.efi ||      
initrd -n bcd         ${base-url}/boot/bcd  bcd||
initrd -n boot.sdi    ${base-url}/boot/boot.sdi  boot.sdi||
initrd -n boot.wim    ${base-url}/sources/boot.wim boot.wim
boot

:exit
exit
```

This got it booting for me. @lrdfrd @antonym

**Reply by bawitdaba** [reply]
> ```
> #!ipxe
> 
> set base-url <location-to-folder-containing-extracted-iso>
> 
> :start
> menu Please choose an option:
> item hirens Hirens Boot CD PE x64
> item exit Exit
> choose --default hirens --timeout 5000 selected
> goto ${selected}
> 
> :hirens
> echo Booting Hirens Boot CD PE x64...
> imgfree
> kernel https://boot.netboot.xyz/wimboot
> initrd -n bootmgr ${base-url}/BOOTMGR       bootmgr ||
> initrd -n bootmgr.efi ${base-url}/bootmgr.efi   bootmgr.efi ||      
> initrd -n bcd         ${base-url}/boot/bcd  bcd||
> initrd -n boot.sdi    ${base-url}/boot/boot.sdi  boot.sdi||
> initrd -n boot.wim    ${base-url}/sources/boot.wim boot.wim
> boot
> 
> :exit
> exit
> ```
> 
> This got it booting for me. @lrdfrd @antonym

While this does work, it seems like only the contents of boot.wim is loaded. None of the programs from the iso root "\Programs" are available which is the majority of the programs in the boot cd.

Edit:
After researching this it looks like Hirens PE creates a Y: drive with the programs that it loads via pecmd.ini in the system32 folder.

I was able to fix this and restore the Y: drive when network pxe booting.

I created a small powershell script "mount_iso.ps1" to mount the ISO from a network path to the Y: drive

> $IsoPath = "\\networkpath\Software\Netboot\asset-mirror\hbcd\HBCD_PE_x64.iso"
> 
> $isoDrive = Mount-DiskImage `
>     -ImagePath $IsoPath `
>     -StorageType ISO `
>     -PassThru
> 
> $isoMountLetter = ($isoDrive | Get-Volume).DriveLetter
> 
> if ($isoMountLetter -ne "Y") {
>     Set-Partition -DriveLetter $isoMountLetter -NewDriveLetter Y
> }

I then modified the pecmd.ini to add this line
`	EXEC =!%WinDir%\System32\WindowsPowerShell\v1.0\powershell.exe -NoProfile -ExecutionPolicy Bypass -File "%WinDir%\System32\mount_iso.ps1"
`
and that is added to the PreShell function of pecmd.ini on line 4

> _SUB PreShell
> 	TEXT
> 	TEXT Initializing Hiren's BootCD PE...\n> Performing Additional Tasks L59 T39 $20
> 	EXEC !=%WinDir%\System32\winpeshl.exe
> 	EXEC =!%WinDir%\System32\WindowsPowerShell\v1.0\powershell.exe -NoProfile -ExecutionPolicy Bypass -File "%WinDir%\System32\mount_iso.ps1"
> 	EXEC =@!%WinDir%\System32\LetterSwap.exe /auto /SetLetter Y:\HBCD_PE.ini /Log %Temp%\LetterSwap.log
> 	EXEC @!%ProgramFiles%\Y_Drive_Checker.bat
> 	EXEC @!%WinDir%\System32\FixScreen.exe
> 
> 	// Load Users Drivers from Y before shell in background (Uncommented By User Option)
> 	TEXT
> 	TEXT Initializing Hiren's BootCD PE...\n> Installing User Custom Drivers L59 T39 $20
> 	EXEC @!%WinDir%\System32\pnputil.exe /add-driver Y:\CustomDrivers\*.inf /subdirs /install
> 	TEXT
> 	TEXT Initializing Hiren's BootCD PE...\n> Almost Done L59 T39 $20
> 
> 	// Load external Autorun.cmd or PecmdStartup.ini to run program, Install external drivers, keyboard,...
> 	IFEX Y:\XPEStartup.ini,LOAD Y:\XPEStartup.ini
> 	IFEX Y:\XPEStartup.cmd,EXEC @!Y:\XPEStartup.cmd
> 
> 	INIT IU,3000
> _END

I then updated my pxe in netboot.xyz to inject these files at boot. These files get inject to the system32 folder and insert/overwrite existing files so you do not need to mess with modifying boot.wim

> #!ipxe
> 
> set hbcd-url ${assets_base_url}/asset-mirror/hbcd/extracted
> 
> :start
> menu Please choose an option:
> item hbcd Hirens Boot CD PE x64
> item exit Exit
> choose --default hbcd --timeout 5000 selected
> goto ${selected}
> 
> :hbcd
> echo Booting Hirens Boot CD PE x64...
> imgfree
> kernel http://${boot_domain}/wimboot gui
> initrd -n mount_iso.ps1       ${assets_base_url}/asset-mirror/hbcd/mount_iso.ps1                 mount_iso.ps1 ||
> initrd -n pecmd.ini           ${assets_base_url}/asset-mirror/hbcd/pecmd.ini                     pecmd.ini ||
> initrd -n bootmgr             ${hbcd-url}/BOOTMGR                                                bootmgr ||
> initrd -n bootmgr.efi         ${hbcd-url}/bootmgr.efi                                            bootmgr.efi ||      
> initrd -n bcd                 ${hbcd-url}/boot/bcd                                               bcd||
> initrd -n boot.sdi            ${hbcd-url}/boot/boot.sdi                                          boot.sdi||
> initrd -n boot.wim            ${hbcd-url}/sources/boot.wim                                       boot.wim
> boot
> 
> :exit
> exit
